### DISCUSSION TITLE
Storage Cost Estimates

### DISCUSSION THREAD

**Discussion by jturan** [CMSgov/price-transparency-guide#147] [General]
Given that we now have less than six months to generate these files, and coupled with the fact that we're still waiting for technical guidance on numerous questions and discrepancies on schema design, duplicative data, and duplicative files, I'd like to start a consolidated discussion on estimated storage costs to host them.

Over the last few weeks, I've seen estimates from various network originators that their in-network machine-readable files (MRFs) will be at least 1TB in a zipped format. Using that baseline info relative to the schema requirements of one file per plan, here's a hypothetical breakdown of a third-party administrator (TPA) that is expected to host these files for 150 self-insured employers.  

Cost Estimate Breakdown:
- Using Azure for cloud-hosting, 1TB of geo-redundant storage is $185 per month per TB. Other cloud-hosting providers have similar pricing.
- **A 1TB file would cost $2,200 per year to host. Not a huge cost, but when multiplied by 150, that balloons to $330,000 to host just the in-network files for all employers. Most of these files will contain 99%+ duplicative data.**
- These costs are astronomical, but many benefits administrators in the country have far more than 150 employers -- many of them have thousands of employers. **Do the quick math, and you'll see the annual storage cost for the largest administrators is now in the millions.**

To the CMS team, I understand you likely don't have the approval yet to respond to this category of questions, especially with how fast they're coming in, but we're getting to a critical point where these storage costs will escalate to entirely unreasonable levels if some of these suggestions aren't implemented.

**To the community, I would encourage you share similar cost estimates below based on how and where you are storing the data. I realize much of this info might be sensitive to share, but the more specific we can be with the tangible costs to store this data, the better the content the CMS team can take back to their peers for review.**

**Comment by jeffrey-price** [comment]
@jturan, thank you for reiterating this concern. We've not run the full set of estimates, but at a high level, our storage demands are shaping to be significant and your concerns echo our own. We will have hundreds of clients each with different combinations of leased and domestic networks and OON/PBM products (although your topic doesn't elaborate on these file types, this only further makes the point).

I'd also add that it's not just storage costs raising concern here. There's also the matter of data transfer fees. Given the intent is that these be made available to the public (e.g. anonymous users, bots, etc) and the fact that we have limited ability to restrict much of anything except the most obvious abuse... Using your example above, even if we were allowed to restrict the users/consumers and estimate that all 3 files for each of the 150 clients would ONLY be downloaded 1x / month, we're talking quite nearly another $100k in data transfer annually (AWS). Again, many will have far more clients than that and the files will be downloaded by far more "users" and/or far more frequently than 1x / month. Data transfer too could easily be $1MM+.

As an aside, is there anything in the ruling that explicitly permits these files to be compressed? One would assume it to be acceptable out of sheer necessity, but the intent and nature of the files would seem to inveigh against the use of file-based compression (one would also assume that HTTP compression would be allowed and an absolute must).

**Reply by ghost** [reply]
@pricejt75 Please see the question, can the files be zipped? #59 The CMS moderator @shaselton-usds states, "There are no restrictions in zipping the files.". Our organization and mostly all those I have been in contact with plan on compressing their files when they are publicly posted.

**Reply by jeffrey-price** [reply]
@BobSyracuse. I'm sure I'd read #59 before but I did not recall having seen the authoritative response. Thank you for the reference/clarification.

**Comment by jonberkery-work** [comment]
@jturan, agree this size of this data is getting out of hand.  We are about 25% through the development of this entire process.  We are forecasting an **uncompressed** size for 1 set of files (in-network, allowed amount, pharmacy) of 6TBs.  Multiply that out by the number of employers we insure (roughly 2700) and you get an unimaginable number.  As you can see this far surpasses the estimate of 5GBs quoted in the rule.

Since much of this data is the same across employers, we are considering dynamically generating the "header" information on the file (i.e. - reporting entity name, reporting entity type, plan name, plan id type, plan id, plan market, etc.).  While this reduces the storage required to host these files, it does not reduce the size of the data that will be transmitted to the "file consumer."

Thanks again for continuing to bring this up, I agree it is frustrating - there are alot of good ideas being presented in this forum, but there is never any final decision made.  Frustrating and confusing.

**Reply by Devyandu** [reply]
> @jturan, agree this size of this data is getting out of hand. We are about 25% through the development of this entire process. We are forecasting an **uncompressed** size for 1 set of files (in-network, allowed amount, pharmacy) of 6TBs. Multiply that out by the number of employers we insure (roughly 2700) and you get an unimaginable number. As you can see this far surpasses the estimate of 5GBs quoted in the rule.
> 
> Since much of this data is the same across employers, we are considering dynamically generating the "header" information on the file (i.e. - reporting entity name, reporting entity type, plan name, plan id type, plan id, plan market, etc.). While this reduces the storage required to host these files, it does not reduce the size of the data that will be transmitted to the "file consumer."
> 
> Thanks again for continuing to bring this up, I agree it is frustrating - there are alot of good ideas being presented in this forum, but there is never any final decision made. Frustrating and confusing.

Hi @jonberkery-work 

I have a question for you if you do not mind answering. 
I am assuming you will host only one file if there are say, 30 flavors of the same plan with identical content just plan id, plan name is different. Your file listing will have all 30 plans. Right ?

For a 5 gig zipped file how will you swap out part of the content on-the-fly when somebody is downloading a copy? 
Thanks
DB

**Reply by jonberkery-work** [reply]
Hey @Devyandu .... good questions, and ones we have been mulling over for some time now.

To your question .... "I am assuming you will host only one file if there are say, 30 flavors of the same plan with identical content just plan id, plan name is different. Your file listing will have all 30 plans. Right ?" - Since the plan-id field is defined as a single string, and not an array of strings, we feel we can't list all 30 plans in the plan-id field.  Also, what would you put in the plan-name field if there were 30 different plan-ids.  Also the rule says self-funded employers are responsible for their file (we have elected to build the file for them), they are not going to be happy if their file is a file that has other customer ids in it also.

For a simple description of our design, we are planning to 
1. Store "incomplete" files (specifically - files that contain only the X number off unique in-network arrays we have).
2. Create a web page with links (estimates look to be 8,100 links - (3*2,700)) that represent the required file naming structure.
3.  When a link is click - a web application will interpret the request (the request will have plan-name in it), look up the required header information based off the submitted plan-name (plan-id-type, plan-id, plan-market), build that header, append that header to the appropriate in-network array, and start streaming that now "complete" file to the requestor.

That is the best approach we could come up with, that met the Rule as closely as possible.  I am all ears for a better approach that helps both the host and the file consumer.

**Reply by Devyandu** [reply]
Hi @jonberkery-work 

#1 and #2 are very similar to the approach that we wish to achieve, but, I am not yet sure of the technology for the #3 bullet you mention. 
If the "incomplete file" is in a storage account (hopefully zipped to reduce storage cost), then this web app would need to unzip a huge file, do some string replacements on the fly, zip the file back to reduce egress cost.

What web technology are you planning which can achieve this on files which are potentially gigabytes in size? I am worried that will be a lot of processing when a file link is clicked.

Any thoughts or pointers are appreciated.

**Comment by garysolebc** [comment]
We have very similar storage, transfer and cost concerns.  We are estimating 3.2TB uncompressed and roughly 3-4 hundred INN files.  That does not include the Rx files that will also be very large.  Since we have to wait for files from associates, we only have 20 days to create and transmit 1-2 Petabytes of data.  That is just us - multiply by all payers, do that monthly and maybe we break the Internet :).  Also, we are struggling a lot with wanting to provide accurate prices in a static space that requires dynamic context.  A standards based API approach may be a much better solution.
We may need to have a concerted group effort to change the solution approach.

**Reply by ericsilvertx** [reply]
Would you be willing to explain how you have come up with the estimate of 3.2 TB.

I am assuming it is mostly because you have many payers and many plans to extract for.

**Reply by garysolebc** [reply]
I am using a spreadsheet calculator created by another
A lot of assumptions on our side but:
10,000 TINs, 
25000 billing codes
2 average service codes per price
1 average rate type
400 bytes per element.
There is also estimated byte sizes for the Negotiated Price Object, Provider Array etc.

Uncompressed, but that should be the state of the final file for consumption.
I am assuming that other payers we are combining with have similar size to us.  
This estimate is also at the "network' level, it could be higher when combining networks for a product/plan.

**Reply by vijayanandramamoorthy** [reply]
Hello, I have seen that spreadsheet. In our opinion that spreadsheet under estimates the file size because of the assumption of 2 service codes per price. Based on our analysis, over 85% of the codes apply to most of the locations listed in the CMS Place of Service Code Set. So our conclusion is, 40 average service codes per price is a realistic assumption based on the current schema. The Internet might survive but I don't think our websites will :)

**Reply by ghost** [reply]
@garysolebc @jonberkery-work Gary, I think the number of "other payers" we will have to combine which Jon B has done a rough estimate will be much more. I think the other payers we will need to include will cause us to append additional information in the TBs range pre-compression. Also, I think the number of billing codes default value in the spreadsheet could be off a bit based on what some of the members of my business team are telling me for our CPT and HCPC count.

**Reply by ericsilvertx** [reply]
Can someone points out to where we can get access to the Excel Model for Sizing please

**Reply by Paul-Sousa** [reply]
If the mystery spreadsheet is the one I think it is, it also is likely underestimating a good amount as it doesn't account for Bundle Code Objects or Covered Service Objects.

**Comment by ericsilvertx** [comment]
Could you upload the spreadsheet or share it from perhaps another
location.  The size is a factor we are trying to get a handle on

**Reply by ghost** [reply]
@ericsilvertx Unfortunately it was developed for in-house use and can not be shared outside the organization.

**Comment by ericsilvertx** [comment]
Is anyone willing to share the spreadsheet?

**Reply by Devyandu** [reply]
I am not the original poster but this is the file I made for my use. Hope it helps.
https://github.com/Devyandu/price-transparency-rate-file-estimator-excel

It is primitive but it can provide you a rough order if magnitude of size.

**Comment by jturan** [comment]
This week, the US Chamber of Commerce [filed a lawsuit](https://www.bloomberglaw.com/public/desktop/document/ChamberofCommerceoftheUnitedStatesofAmericaetalvUnitedStatesDepar/1?1628775500) against HHS over certain aspects of the transparency rule. They called out the "skewed and defective cost-benefit analysis in considering and justifying the machine-readable files requirement". TBD on if this will have an impact on the requirements and format of the files related to the concerns we've been raising, but it's definitely worth keeping an eye on. Below are specific sections of the complaint relevant to this community:

_108. This shortcoming is compounded by Defendants’ decision to use a skewed and defective cost-benefit analysis in considering and justifying the machine-readable files requirement. As noted above, Defendants estimated that the quantified costs of the Rule will exceed $3.5 billion in just the first year of implementation. See supra, ¶ 64. Yet the Rule neglected to quantify more than a dozen additional costs that would have greatly increased this massive number even further, including the “[p]otential increase in health care costs if consumers confuse cost with quality and value of service,” the “[p]otential increase in cyber security costs . . . to prevent data breaches,” the cost “to conduct quality control reviews of the information” required to be disclosed in the machine-readable files, and the cost to “renegotiate contracts in order to remove gag clauses” that currently protect commercially sensitive pricing information from disclosure. 85 Fed. Reg. at 72,260. There is no obvious reason why (unlike the costs that Defendants did quantify) all of these costs are incapable of being estimated. Defendants’ failure to attempt even a rough estimate of these and other potentially significant costs, coupled with the inadequate estimates that Defendants did prepare, is further evidence of an agency decision making process that was arbitrary and capricious._

_109. Finally, Defendants acted arbitrarily and capriciously in refusing to extend the deadline for implementing the machine-readable files. The machine-readable files present a logistical and technological challenge of enormous size and complexity. Thus, it is no surprise that a “majority of commenters strongly recommended delaying the proposed applicability date” of January 1, 2022. Id. at 72,252. Defendants dismissed concerns about this rushed timeline out of hand, asserting that the process would be “straightforward” because the files involve “already existing data.” Id. at 72,253. But whether the data exist—in some form, somewhere in the universe—is beside the point. Plans and issuers must develop new systems for gathering thousands upon thousands of disparate data points (which, to the extent they are not in a plan’s or issuer’s possession, are likely to be particularly time-consuming and costly to obtain), processing them, and converting them into the machine-readable format. That is no “straightforward” undertaking for even the largest entities subject to the Rule—and will likely be even more problematic for smaller entities, including self-insured employers such as the U.S. Chamber._

**Reply by jturan** [reply]
@BobSyracuse given your level of contribution to these discussions and familiarity across the various challenges we're facing and how to potentially resolve them, I'd like to have a separate email discussion (if you're open to it) about some documentation I'm working on for an upcoming listening session held by HHS to discuss this very topic. I'm not actually participating in the session, but rather working with a contact that represents an agency that has a direct relationship with HHS and has been relaying many of our concerns behind the scenes. Assuming the session is scheduled, this person will very likely receive an invite, and has asked me for some additional details to have on-hand.

If you're interested in chatting, please send me an email to jturan@hey.com. And to the rest of the community, I will keep you updated on those efforts if they make any meaningful progress.

**Reply by ghost** [reply]
@jturan I will contact you. I shared your post yesterday about the court filing within my organization and they were very interested in seeing it.

**Reply by jeffrey-price** [reply]
I'm sure others have seen this already, but the Pharmaceutical Care Management Association (PCMA) has also [filed suit](https://www.pcmanet.org/wp-content/uploads/2021/08/2021-08-12-1-PCMA-v.-HHS-Complaint.pdf) this week challenging parts of the price transparency rule:

> 2. The Machine-Readable File Requirement Is Arbitrary And Capricious 

> 173. The machine-readable file requirement is also arbitrary and capricious for three
> independent reasons: It will not achieve the Rule’s intended purposes, the Departments’ costbenefit analysis is fatally flawed, and the Departments imposed an unrealistically short implementation timeline for the machine-readable files. 

and

> ii. The Cost-Benefit Analysis Is Fatally Flawed
> 
> 177. The Departments also acted unreasonably by failing to adequately quantify and
> assess the costs and benefits of the machine-readable file requirement. See, e.g., Sorenson
> Commc’ns Inc. v. FCC, 755 F.3d 702, 708-09 (D.C. Cir. 2014) (holding that agency action based
> on “sheer speculation” rather than “evidence” was arbitrary and capricious). As the U.S.
> Chamber of Commerce noted, the Proposed Rule itself listed some “10 specific cost elements
> that it did not attempt to quantify,” failed to account for some costs altogether, and “grossly
> underestimated” other costs while failing to quantify offsetting benefits. U.S. Chamber
> Comment Letter at 7-8 (Jan. 29, 2020), www.regulations.gov/comment/CMS-2019-0163-19418.
> 
> 178. The final rule similarly failed to quantify relevant costs and benefits. The Rule
> identifies 14 specific costs that the Departments never attempted to quantify, including the
> “increase in cyber security costs . . . to prevent data breaches,” the “increase in health care costs
> if consumers confuse cost with quality and value of service,” the cost “to conduct quality control
> reviews of the information” required to be disclosed in the machine-readable files, the costs of
> “renegotiat[ing] contracts in order to remove gag clauses in order” to disclose the information
> required by the final rule, and the “increase in costs to consumers and issuers if providers or 
> prescription drug manufacturers engage in anticompetitive behaviors.” Transparency Rule, 85
> Fed. Reg. at 72260. Because these costs could be significant, the Departments’ failure even to
> attempt to quantify them is arbitrary and capricious.
